drivers/soundwire/intel_init.c

```diff
@@ -363,6 +364,25 @@ void sdw_intel_exit(struct sdw_intel_ctx *ctx)
 }
 EXPORT_SYMBOL_NS(sdw_intel_exit, SOUNDWIRE_INTEL_INIT);
 
+void sdw_intel_pre_exit(struct sdw_intel_ctx *ctx)
+{
+	struct sdw_intel_link_res *link;
+
+	list_for_each_entry(link, &ctx->link_list, list) {
+		struct sdw_bus *bus = &link->cdns->bus;
+		struct sdw_slave *slave;
+
+		list_for_each_entry(slave, &bus->slaves, node) {
+			/*
+			 * Make sure peripherals are fully resumed and don't need to access
+			 * SDW registers any more.
+			 */
+			pm_runtime_resume(&slave->dev);
```

Comment:
this sounds fine but we need to explain why errors are ignored, and I wonder if for consistency with the prepare step we should use the macro device_for_each_child()

Reply:
> this sounds fine but we need to explain why errors are ignored, and I wonder if for consistency with the prepare step we should use the macro device_for_each_child()

@plbossart Can we change [pm_request_resume(dev);](https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/intel_auxdevice.c#L457) to `pm_runtime_resume(dev);` and reuse `intel_resume_child_device()`?

Reply:
I think so @bardliao. I don't see any reason why the .prepare should rely on an async behavior before stopping the clock, that's also risky with a codec driver potentially doing stuff after the clock is stopped.

That would need to be two patches though, so that we can bisect the changes from pm_request_resume() to pm_runtime_resume() separately from the pre-exit stuff for the remove operation.